---
title: "Introduction to crisprViz"
author: 
- name: Luke Hoberecht
  affiliation: Department of Data Science and Statistical Computing, gRED, Genentech
  email: hoberecl@gene.com
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc_float: true
    #theme: paper
    number_sections: true
vignette: >
  %\VignetteIndexEntry{Introduction to crisprViz}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r, include=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE)
```


# Introduction

The `crisprViz` package enables the graphical interpretation of `GuideSet` objects from the `crisprDesign` package by plotting guide RNA (gRNA) cutting locations against their target gene or other genomic region. These genomic plots are constructed using the `Gviz` package from Bioconductor.

This vignette walks through several use cases that demonstrate the range of and how to use plotting functions in the `crisprViz` package. This vignette also makes heavy use of the `crisprDesign` package to manipulate `GuideSet` objects in conjunction with plotting in the process of gRNA design. For more information about the `crisprDesign` package see [vignettes].


# Installation

`crisprViz` can be installed from Bioconductor using the following commands in an R session:

```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("crisprViz")
```

# Use cases

All examples in this vignette will use human genome assembly `hg38` from the `BSgenome.Hsapiens.UCSC.hg38` package and gene model coordinates from Ensembl release 104 via the `crisprDesignData` package.

```{r, message=FALSE, warning=FALSE}
library(BSgenome.Hsapiens.UCSC.hg38)
library(crisprDesign)
library(crisprDesignData)
data(txdb_human, package="crisprDesignData")
library(crisprViz)
```

## Visualizing the best gRNAs for a given gene

Suppose we want to design the four best gRNAs using the SpCas9 CRISPR nuclease to knockout the human KRAS gene. To have the greatest impact on gene function we want to prioritize gRNAs that have greater isoform coverage and target closer to the 5' end of the CDS.

We will start by finding all possible gRNAs that target the CDS of KRAS.

``` {r}
kras <- queryTxObject(txdb_human,
                      featureType="cds",
                      queryColumn="gene_symbol",
                      queryValue="KRAS")
gs <- findSpacers(kras,
                  crisprNuclease=SpCas9,
                  bsgenome=BSgenome.Hsapiens.UCSC.hg38)
gs <- unique(gs)
length(gs) # number of candidate gRNAs
```

Before we plot all of our candidate gRNAs, let's first generate a simple plot with a few gRNAs to familiarize ourselves with some plot components and options.

```{r}
plotGuideSet(gs[1:4],
             geneModel=txdb_human,
             targetGene="KRAS")
```
There are a few things to note here.

 * The ideogram track and genome axis track are at the top of our plot and give us coordinate information.
 * Our `targetGene` KRAS is plotted next, using coordinates from the provided `geneModel`, followed by our spacer subset. The name of each track is given on the left.
 * The strand information for each track is included in the label: `<-` for reverse strand and `->` for forward strand.
 * While we can identify which exon each spacer targets (which may be sufficient), the plot window is too large to provide further information.
 * The plot only shows the 3' end of KRAS, rather than the entire gene.
 
This last point is important: the default plot window is set by the spacers' ranges in the input `GuideSet` object. We can manually adjust this window by using the `from`, `to`, `extend.left`, and `extend.right` arguments. Here is the same plot adjusted to show the whole KRAS gene, which also reveals an additional isoform that is not targeted by any spacer in this example subset.
 
```{r}
kras_tx <- queryTxObject(txdb_human,
                         featureType="transcripts",
                         queryColumn="gene_symbol",
                         queryValue="KRAS")
from <- min(start(kras_tx))
to <- max(end(kras_tx))
plotGuideSet(gs[1:4],
             geneModel=txdb_human,
             targetGene="KRAS",
             from=from,
             to=to,
             extend.left=1000,
             extend.right=1000)
```
 
As calculated above, there are a total of `r length(gs)` candidate gRNAs targeting the CDS of KRAS. Including all of them could crowd the plot space, making it difficult to interpret. To alleviate this we can hide the gRNA labels by setting the `showGuideLabels` argument to `FALSE`.

``` {r fig.height=10}
plotGuideSet(gs,
             geneModel=txdb_human,
             targetGene="KRAS",
             showGuideLabels=FALSE,
             from=from,
             to=to,
             extend.left=1000,
             extend.right=1000)
```

At the gene level, the plot window is too large to discern details for each spacer target. However, we can see five distinct clusters of spacer target locations that cover the CDS of KRAS. The spacers in the 5'-most cluster (on the reverse strand) target the only coding region of the gene that is expressed by all isoforms, making it an ideal target for our scenario.

We can see which gRNAs target this region by returning `showGuideLabels` to its default value of `TRUE`, and by adjusting the plot window to focus on our exon of interest.

```{r}
# new window range around target exon
exon_ranks <- mcols(kras)$exon_rank
min_exon_rank <- min(exon_ranks)
targetRegion <- kras[exon_ranks == min_exon_rank]
from <- min(start(targetRegion))
to <- max(end(targetRegion))
plotGuideSet(gs,
             geneModel=txdb_human,
             targetGene="KRAS",
             from=from,
             to=to,
             extend.left=20,
             extend.right=20)
```

At this resolution we can get a much better idea of spacer location and orientation. In particular, the PAM sequence is visible as a narrow box on the 3' side of our protospacer sequences. We can also distinctly see which spacer targets overlap--it may be best to avoid pairing such spacers in some applications lest they sterically interfere with each other.

If we have many gRNA targets in a smaller window and are not concerned with overlaps, we can configure the plot to only show the `pam_site`, rather than the entire protospacer and PAM sequence, by setting `pamSiteOnly` to `TRUE`.

```{r}
plotGuideSet(gs,
             geneModel=txdb_human,
             targetGene="KRAS",
             from=from,
             to=to,
             extend.left=20,
             extend.right=20,
             pamSiteOnly=TRUE)
```

Let's filter our `GuideSet` by the spacer names in the plot then briefly assess gRNA quality by adding an on-target score. We can pass the added score column to `onTargetScores` to color the spacers according to that score, with darker blue colors indicating higher scores. Note that for this plot we need not provide values for `from` and `to`, as the plot window adjusts to our filtered `GuideSet`.

```{r, message=FALSE, warning=FALSE}
goodGuides <- c("spacer_80", "spacer_84", "spacer_88", "spacer_92",
                "spacer_96", "spacer_100", "spacer_104", "spacer_108",
                "spacer_112")
gs <- gs[goodGuides]

gs <- addOnTargetScores(gs,
                        methods="deephf")
plotGuideSet(gs,
             geneModel=txdb_human,
             targetGene="KRAS",
             onTargetScore="score_deephf")
```

<!-- ![legend caption](file path) -->


## Plotting for precision targeting

For a given CRISPR application, the target region may consist of only several base pairs rather than an exon or entire gene CDS. In these instances it is crucial to know exactly where the gRNAs target, and plots of gRNAs must be at a resolution capable of distinguishing individual bases. This is often the case for CRISPR base editor (CRISPRbe) applications, as the editing window for each gRNA is narrow and the results are specific to each target sequence.

In this example we will zoom in on a few gRNAs targeting the 5' end of the human GPR21 gene. We want our plot to include genomic sequence information so we will set the `bsgenome` argument to the same `BSgenome` object we used to create our `GuideSet`.

```{r, fig.height=6}
gpr21 <- queryTxObject(txdb_human,
                       featureType="cds",
                       queryColumn="gene_symbol",
                       queryValue="GPR21")
be <- findSpacers(gpr21,
                  crisprNuclease=BE4max,
                  bsgenome=BSgenome.Hsapiens.UCSC.hg38)
plotGuideSet(be[1:4],
             geneModel=txdb_human,
             targetGene="GPR21",
             bsgenome=BSgenome.Hsapiens.UCSC.hg38,
             from=123034550,
             to=123034610)
```

The genomic sequence is given at the bottom of the plot as color-coded boxes. The color scheme used to describe the nucleotides is given in the biovizBase (link) package. If the plot has sufficient space, it will display nucleotide symbols rather than boxes. We can accomplish this by plotting a narrower range or by increasing the width of our plot space (see "Setting plot size" section).

The plot above was generated with a plot space width of 7 inches; here's the same plot after we increase the width to 10 inches.

```{r, fig.width=10, fig.height=6}
# increase plot width from 7" to 10"
plotGuideSet(be[1:4],
             geneModel=txdb_human,
             targetGene="GPR21",
             bsgenome=BSgenome.Hsapiens.UCSC.hg38,
             from=123034550,
             to=123034610)
```


## CRISPRa and adding genomic annotations

In this scenario we want to express the human MMP7 gene via CRISPR activation (CRISPRa). We will use the SpCas9 CRISPR nuclease and obtain TSS coordinates from the `crisprDesignData` package and begin our search for candidate gRNAs in the 2kb window immediately upstream of the TSS. To aid in the selection process, we will include additional genomic annotation tracks which identify either desirable or undesirable targets in our region of interest.

Let's begin by constructing and plotting our `GuideSet`, and adding a track of repeat elements using the `annotations` argument.

```{r}
data("tss_human", package="crisprDesignData")
data("gr.repeats.hg38", package="crisprDesignData")
mmp7 <- queryTss(tss_human,
                 queryColumn="gene_symbol",
                 queryValue="MMP7",
                 tss_window=c(-2000, 0))
gs <- findSpacers(mmp7,
                  crisprNuclease=SpCas9,
                  bsgenome=BSgenome.Hsapiens.UCSC.hg38)
plotGuideSet(gs,
             geneModel=txdb_human,
             targetGene="MMP7",
             guideStacking="dense",
             annotations=list(Repeats=gr.repeats.hg38),
             pamSiteOnly=TRUE,
             from=start(mmp7),
             to=end(mmp7),
             extend.left=600,
             extend.right=100)
```

Some of our candidate gRNAs target repeat elements, and likely have many off-targets as a result. Let's remove these gRNAs and regenerate the plot.

```{r}
gs <- removeRepeats(gs,
                    gr.repeats=gr.repeats.hg38)
plotGuideSet(gs,
             geneModel=txdb_human,
             targetGene="MMP7",
             guideStacking="dense",
             annotations=list(Repeats=gr.repeats.hg38),
             pamSiteOnly=TRUE,
             from=start(mmp7),
             to=end(mmp7),
             extend.left=600,
             extend.right=100)
```

We can similarly filter out gRNAs that overlap SNPs. However, rather than adding SNPs as a separate annotation, we can incorporate SNPs into our `GuideSet` with `addSNPAnntation()` and plot the SNPs track by having `includeSNPTrack` as `TRUE`. Note that since SNPs are obtained from the `GuideSet`, only SNPs that overlap with gRNAs in the `GuideSet` will be plotted. The `addSNPAnnotation()` function also appends a logical `hasSNP` column, which we could use to select gRNAs that do not overlap SNPs.

Conversely, there are specific genomic regions that would be beneficial to target, such as CAGE peaks and DNase I Hypersensitivity tracks (briefly, why). Let's retrieve those annotations from the Bioconductor package `AnnotationHub` and plot them alongside our gRNAs.

```{r}
library(AnnotationHub)
ah <- AnnotationHub()

## lift over annotation tracks from hg19
chain <- ah[["AH14150"]]
cage <- ah[["AH5084"]]
cage <- rtracklayer::liftOver(cage, chain)
cage <- unlist(cage)
dnase <- ah[['AH30743']]
dnase <- rtracklayer::liftOver(dnase, chain)
dnase <- unlist(dnase)

plotGuideSet(gs,
             geneModel=txdb_human,
             targetGene="MMP7",
             guideStacking="dense",
             annotations=list(CAGE=cage, DNase=dnase),
             pamSiteOnly=TRUE,
             from=start(mmp7),
             to=end(mmp7),
             extend.left=600,
             extend.right=100)
```

Let's filter our `GuideSet` for guides overlapping the plotted DNase site then regenerate the plot.

```{r}
# filter GuideSet for gRNAs overlapping DNase track
overlaps <- findOverlaps(gs, dnase, ignore.strand=TRUE)
gs <- gs[queryHits(overlaps)]

plotGuideSet(gs,
             geneModel=txdb_human,
             targetGene="MMP7",
             guideStacking="dense",
             annotations=list(CAGE=cage, DNase=dnase),
             pamSiteOnly=TRUE,
             margin=0.4)
```

## Comparing multiple GuideSets targeting the same region

Choice of CRISPR nuclease can be influenced by the abundance of qualify PAM sequences in the target region. For example, we would expect AT-rich regions to have fewer possible targets for the SpCas9 nuclease, whose PAM is NGG. In these regions, the CRISPR nuclease AsCas12a, whose PAM is TTTV, may prove more appropriate. Given multiple `GuideSet`s targeting the same region, we can compare the gRNAs of each in the same plot using `plotMultipleGuideSets`.

Here, we pass our `GuideSet`s targeting an exon in the human gene LTN1 in a named list. Note that there are no available options for displaying guide labels or guide stacking, and only the PAM sites are plotted. We will also add a track to monitor the percent GC content (using a window roughly the length of our protospacers). Not surprisingly, this AT-rich region has fewer targets for SpCas9 compared to AsCas12a.

```{r}
target <- queryTxObject(txdb_human,
                        featureType="cds",
                        queryColumn="exon_id",
                        queryValue="ENSE00001705186")
## SpCas9
gs_cas9 <- findSpacers(target,
                       crisprNuclease=SpCas9,
                       bsgenome=BSgenome.Hsapiens.UCSC.hg38)
gs_cas9 <- unique(gs_cas9)
## AsCas12a
gs_cas12a <- findSpacers(target,
                         crisprNuclease=AsCas12a,
                         bsgenome=BSgenome.Hsapiens.UCSC.hg38)
gs_cas12a <- unique(gs_cas12a)
## plot GuideSets
plotMultipleGuideSets(list(SpCas9=gs_cas9, AsCas12a=gs_cas12a),
                      geneModel=txdb_human,
                      targetGene="LTN1",
                      bsgenome=BSgenome.Hsapiens.UCSC.hg38,
                      margin=0.2,
                      gcWindow=10)
```



# Setting plot size

Plots with many gene isoforms and/or gRNAs may require more space to render
than is allotted by your graphical device's default settings, resulting in an
error. One solution, depending on your graphical device, is offered by the
[**grDevices** package](https://www.rdocumentation.org/packages/grDevices/versions/3.6.2/topics/Devices).

Here is an example using macOS Quartz device.

```{r, eval=FALSE}
grDevices::quartz("Example plot", width=6, height=7)
# plot function
```
